Userland/Libraries/LibVideo/PlaybackManager.h

```diff
@@ -30,18 +30,57 @@ enum class PlaybackStatus {
     Buffering,
     Seeking,
     Stopped,
+    Corrupted,
 };
 
 struct FrameQueueItem {
-    static FrameQueueItem eos_marker()
+    enum class Type {
+        Frame,
+        Error,
+    };
+
+    static FrameQueueItem frame(RefPtr<Gfx::Bitmap> bitmap, Time timestamp)
+    {
+        return FrameQueueItem(move(bitmap), timestamp);
+    }
+
+    static FrameQueueItem error_marker(DecoderError&& error)
+    {
+        return FrameQueueItem(move(error));
+    }
+
+    bool is_frame() { return m_data.has<FrameData>(); }
+    RefPtr<Gfx::Bitmap> bitmap() { return m_data.get<FrameData>().bitmap; }
+    Time timestamp() { return m_data.get<FrameData>().timestamp; }
+
+    bool is_error() { return m_data.has<Optional<DecoderError>>(); }
+    DecoderError error() { return m_data.get<Optional<DecoderError>>().release_value(); }
+    DecoderError release_error() { return m_data.get<Optional<DecoderError>>().release_value(); }
+
+    String debug_string()
+    {
+        if (is_error())
+            return error().string_literal();
+        return String::formatted("frame at {}ms", timestamp().to_milliseconds());
+    }
+
+private:
+    struct FrameData {
+        RefPtr<Gfx::Bitmap> bitmap;
+        Time timestamp;
+    };
+
+    FrameQueueItem(RefPtr<Gfx::Bitmap> bitmap, Time timestamp)
+        : m_data(FrameData { move(bitmap), timestamp })
     {
-        return { nullptr, Time::max() };
     }
 
-    RefPtr<Gfx::Bitmap> bitmap;
-    Time timestamp;
+    FrameQueueItem(DecoderError&& error)
+        : m_data(move(error))
+    {
+    }
 
-    bool is_eos_marker() const { return !bitmap; }
+    Variant<FrameData, Optional<DecoderError>> m_data;
```

Comment:
What happens if there's no frame data, and there's no decode error?
can that even happen?

Reply:
Looks like this optional shouldn't even be an optional, you're deref'ing it as if it always holds a value.

Reply:
> What happens if there's no frame data, and there's no decode error? can that even happen?

If there's no data in the video as it's read by the demuxer, then it should send an EndOfStream error for the playback manager to handle.

> Looks like this optional shouldn't even be an optional, you're deref'ing it as if it always holds a value.

I did this so that the DecoderError doesn't have to be copied (since the error could be OOM), since I can't release the value from the FrameQueueItem without copying, I believe. Feel free to correct me if I'm wrong, or suggest a better way to do this. Otherwise, if you prefer, I can just ignore that copy and remove the Optional.

Reply:
As an alternative to Optional here, I discovered AK::Empty that I can use in the Variant to be able to move the decoder error out of the FrameQueueItem and render it invalid. This works a bit more elegantly, hopefully it's good enough. I'll resolve this based on that change.